### PR TITLE
Add List migrations command

### PR DESCRIPTION
**Pull request by goetas**
|      Q       |   A
|------------- | -----------
| Type         | feature
| BC Break     | -
| Fixed issues | https://github.com/doctrine/migrations/issues/734

#### Summary

as suggested in https://github.com/doctrine/migrations/issues/734, this PR adds a dedicated `list` command for listing migrations.

For now, this is just a draft, since is based on https://github.com/doctrine/migrations/pull/898. For a better diff, check https://github.com/doctrine/migrations/pull/908/commits/3f344d2466ddf74c52c2278d5786a5756b48cd92

### CODE REVIEW

**Comment by goetas** [comment]
This is ready to be reviewed

File: lib/Doctrine/Migrations/Tools/Console/Command/ListCommand.php [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+The <info>%command.name%</info> command outputs a list of all available migrations and their status:
+
+    <info>%command.full_name%</info>
+EOT
+            );
+
+        parent::configure();
+    }
+
+    public function execute(InputInterface $input, OutputInterface $output) : ?int
```
Comment (greg0ire) [inline lib/Doctrine/Migrations/Tools/Console/Command/ListCommand.php]: Why not

```suggestion
    public function execute(InputInterface $input, OutputInterface $output) : int
```

File: lib/Doctrine/Migrations/Tools/Console/Command/ListCommand.php [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+use Symfony\Component\Console\Output\OutputInterface;
+use function array_map;
+use function array_merge;
+use function array_unique;
+use function uasort;
+
+/**
+ * The StatusCommand class is responsible for outputting a list of all available migrations and their status.
+ */
+class ListCommand extends DoctrineCommand
```
Comment (greg0ire) [inline lib/Doctrine/Migrations/Tools/Console/Command/ListCommand.php]: Should we make it final?

```suggestion
final class ListCommand extends DoctrineCommand
```

File: lib/Doctrine/Migrations/Tools/Console/Command/ListCommand.php [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+    public function execute(InputInterface $input, OutputInterface $output) : ?int
+    {
+        $storage       = $this->getDependencyFactory()->getMetadataStorage();
+        $migrationRepo = $this->getDependencyFactory()->getMigrationRepository();
+
+        $availableMigrations = $migrationRepo->getMigrations();
+        $executedMigrations  = $storage->getExecutedMigrations();
+
+        $this->showVersions($availableMigrations, $executedMigrations, $output);
```
Comment (greg0ire) [inline lib/Doctrine/Migrations/Tools/Console/Command/ListCommand.php]: How about
```suggestion
        $this->showVersions(
            $this->getDependencyFactory()->getMigrationRepository()->getMigrations(), // available migrations
            $this->getDependencyFactory()->getMetadataStorage()->getExecutedMigrations(), // executed migrations
            $output
        );
```

Fewer variables, no abbreviations, no assignments so no need to align those, no need to remember that we created a `$storage` variable until it is finally used, and you can even make sentences in the comments if you want to clarify things even more

File: lib/Doctrine/Migrations/Tools/Console/Command/ListCommand.php [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+        $table = new Table($output);
+        $table->setHeaders(
+            [
+                [new TableCell('Migration Versions', ['colspan' => 4])],
+                ['Migration', 'Status', 'Migrated At', 'Execution Time', 'Description'],
+            ]
+        );
+
+        foreach ($versions as $version) {
```
Comment (greg0ire) [inline lib/Doctrine/Migrations/Tools/Console/Command/ListCommand.php]:
```suggestion
        foreach ($this->getSortedVersions($availableMigrationsSet, $executedMigrationsSet) as $version) {
```

**Comment by goetas** [comment]
All suggestions have been applied

File: docs/en/reference/custom-configuration.rst [hunk 5]
```diff
@@ -60,7 +60,9 @@ Once you have your custom integration setup, you can modify it to look like the
         new Command\MigrateCommand($dependencyFactory),
         new Command\RollupCommand($dependencyFactory),
         new Command\StatusCommand($dependencyFactory),
-        new Command\VersionCommand($dependencyFactory)
+        new Command\VersionCommand($dependencyFactory),
+        new Command\ListCommand($dependencyFactory),
+        new Command\SyncMetadataCommand($dependencyFactory),
```
[the remark this answers is not in the document]
Comment (greg0ire) [reply docs/en/reference/custom-configuration.rst]: V comes after S :stuck_out_tongue:

File: docs/en/reference/custom-integration.rst [hunk 6]
```diff
@@ -48,10 +48,12 @@ Now place the following code in the ``migrations`` file:
         new Command\ExecuteCommand($dependencyFactory),
         new Command\GenerateCommand($dependencyFactory),
         new Command\LatestCommand($dependencyFactory),
+        new Command\ListCommand($dependencyFactory),
         new Command\MigrateCommand($dependencyFactory),
         new Command\RollupCommand($dependencyFactory),
         new Command\StatusCommand($dependencyFactory),
-        new Command\VersionCommand($dependencyFactory)
+        new Command\VersionCommand($dependencyFactory),
+        new Command\ListCommand($dependencyFactory),
```
Comment (greg0ire) [inline docs/en/reference/custom-integration.rst]: `ListCommand` appears twice.

Comment (goetas) [reply docs/en/reference/custom-integration.rst]: my bad, fixed

**Pull request merged**